src/pytorch_lightning/core/module.py

```diff
@@ -172,6 +173,21 @@ def lr_schedulers(self) -> Optional[Union[LRSchedulerTypeUnion, List[LRScheduler
         # multiple schedulers
         return lr_schedulers
 
+    @property
+    def trainer(self) -> "pl.Trainer":
+        if not self._running_torchscript and self._trainer is None:
+            raise RuntimeError(f"{self.__class__.__qualname__} is not attached to a `Trainer`.")
+        return self._trainer
+
+    @trainer.setter
+    def trainer(self, trainer: "pl.Trainer") -> None:
+        if not isinstance(trainer, weakref.ProxyTypes):
+            trainer = weakref.proxy(trainer)
+        self._trainer = trainer
+        for v in self.children():
+            if isinstance(v, LightningModule):
+                v.trainer = trainer
```

Comment:
```suggestion
        for v in self.children():
            if isinstance(v, LightningModule):
                v.trainer = trainer
        if trainer is not None:
            trainer = weakref.proxy(trainer)
        self._trainer = trainer
       
```

very nit: This seems a bit cleaner to me (as you pass the core trainer reference to the children and they could theoretically do whatever they want with that). Also you spare one check and order shouldn't really matter here.
And IMO it does not make sense to have a weakref of `None`

Reply:
Looks good to me since `weakref.proxy(t) is weakref.proxy(t)` is True